File: js/src/views/Signer/signer.js
// Copyright 2015, 2016 Ethcore (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';

import { Actionbar, Page } from '../../ui';

import LoadingPage from './containers/LoadingPage';
import RequestsPage from './containers/RequestsPage';

import styles from './signer.css';

export class Signer extends Component {
  static propTypes = {
    signer: PropTypes.shape({
      isLoading: PropTypes.bool.isRequired
    }).isRequired
  };

  render () {
    return (
      <div className={ styles.signer }>
        <Actionbar
          title='Trusted Signer' />
        <Page>
          { this.renderPage() }
        </Page>
      </div>
    );
  }

  renderPage () {
    const { isLoading } = this.props.signer;

    if (isLoading) {
      return (
        <LoadingPage />
      );
    }

    return (
      <RequestsPage />
    );
  }
}

function mapStateToProps (state) {
  return state;
}

function mapDispatchToProps (dispatch) {
  return {};
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Signer);

File: ethcore/engine/src/signer.rs
// Copyright 2015-2020 Parity Technologies (UK) Ltd.
// This file is part of Parity Ethereum.

// Parity Ethereum is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity Ethereum is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.

//! A signer used by Engines which need to sign messages.

use ethereum_types::{H256, Address};
use parity_crypto::publickey::{ecies, Public, Signature, KeyPair, Error};

/// Everything that an Engine needs to sign messages.
pub trait EngineSigner: Send + Sync {
	/// Sign a consensus message hash.
	fn sign(&self, hash: H256) -> Result<Signature, Error>;

	/// Signing address
	fn address(&self) -> Address;

	/// Decrypt a message that was encrypted to this signer's key.
	fn decrypt(&self, auth_data: &[u8], cipher: &[u8]) -> Result<Vec<u8>, Error>;

	/// The signer's public key, if available.
	fn public(&self) -> Option<Public>;
}

/// Creates a new `EngineSigner` from given key pair.
pub fn from_keypair(keypair: KeyPair) -> Box<dyn EngineSigner> {
	Box::new(Signer(keypair))
}

struct Signer(KeyPair);

impl EngineSigner for Signer {
	fn sign(&self, hash: H256) -> Result<Signature, Error> {
		parity_crypto::publickey::sign(self.0.secret(), &hash)
	}

	fn decrypt(&self, auth_data: &[u8], cipher: &[u8]) -> Result<Vec<u8>, Error> {
		ecies::decrypt(self.0.secret(), auth_data, cipher)
	}

	fn address(&self) -> Address {
		self.0.address()
	}

	fn public(&self) -> Option<Public> {
		Some(*self.0.public())
	}
}
